added AI (#971)
microsoft/code-with-engineering-playbook: docs/developer-experience/artificial-intelligence.md

> # Artificial Intelligence

Reviewer: I feel the title is misleading, I think it needs to be Copilots and not AI as AI is a big topic 
```suggestion
# Copilots
```

There are a number of AI tools that can improve the developer experience. This article will discuss tooling that is available as well as advice on when it might be appropriate to use such tooling.

## GitHub Copilot

The current version of GitHub Copilot can provide code completion in Visual Studio Code. It is a VS Code extension that can be installed from the VS Code Marketplace. It is currently in beta and requires a GitHub account to use. It is free to use during the beta period. It is currently available for VS Code only, but there are plans to support other IDEs in the future. It is currently available for the following languages: JavaScript, TypeScript, Ruby, Go, Python, C#, C++, and Java. You can find more information about GitHub Copilot [here](https://copilot.github.com/).

> Developers on our team have found the following use-cases for GitHub Copilot:

Reviewer 2: Can we reword this (and similar statements below). This wording fits great in blogs but feels a bit out of place in the playbook. Perhaps something like... "some example use-cases are"

Reviewer: Agreed, the playbook should be more about hard cold facts.

Author: Yes, we can. However, realize that this content comes from the "sharing sprint" for a devcrew. We are not the product group saying facts about what use-cases are supported by GitHub Copilot. We are a devcrew who used the tools in an engagement - it can do a lot of things - these are the things that we as a crew found productive. Our guiding principal for sharing, in fact, has been that we don't share anything that we haven't actually vetted and proven to be useful on our team. If we just say "some example use-cases" then it seems like we are losing this important distinction.

Reviewer 2: By sharing it in the EF playbook, the connotation is already that someone in CSE has found it useful, no need to call it out here.

- __Write Documentation__. For example, the above paragraph was written using Copilot.
- __Write Unit Tests__. Given that setup and assertions are often consistent across unit tests, Copilot tends to be very accurate.
- __Unblock__. It is often hard start writing when staring at a blank page, Copilot can fill the space with something that may or may not be what you ultimately want to do, but it can help get you in the right headspace.

If you want Copilot to write something useful for you, try writing a comment that describes what your code is going to do - it can often take it from there.

## GitHub Copilot Labs

> Copilot has extension that offers additional features that are not yet ready for prime-time. You can install it from the VS Code Marketplace. These features include:

Reviewer 2: ```suggestion
Copilot has an extension that offers additional features that are not yet ready for prime-time. You can install it from the VS Code Marketplace. These features include:
```

Would be good to point to the extension page

Author: done.

- __Explain__. Copilot can explain what the code is doing in natural language.
- __Translate__. Copilot can translate code from one language to another.
- __Brushes__. You can select code that Copilot then modifies inline based on a "brush" you select, for example, to make the code more readable, fix bugs, improve debugging, document, etc.
- __Generate Tests__. Copilot can generate unit tests for your code. Though currently this is limited to JavaScript and TypeScript.

## GitHub Copilot X

> This extention for Copilot is currently in technical preview. It will enable the following additional features:

Reviewer 2: ```suggestion
This extension for Copilot is currently in technical preview. It will enable the following additional features:
```

> This extention for Copilot is currently in technical preview. It will enable the following additional features:

Reviewer 2: If we describe the availability, this page will immediately be outdated when the component is released. To avoid this, change this and say, see ... for availability

Author: I think we may just have to update it once it becomes available. How and where to sign-up and how to check whether you are "in" was really hard for folks. I had several people reach out to me for help, so I put the links here.

Reviewer 2: If there is an issue where people can't tell from the product page how to sign up, we should make an effort to connect with the product group to make this easier.

Author: As discussed, removing this section. I will consider if this is better for the wiki or just point people to the official documentation.

- __Chat__. Rather than just providing code completion, Copilot will be able to have a conversation with you about what you want to do. It has context about the code you are working on and can provide suggestions based on that context.
- __Explain__. Copilot can explain what the code is doing in natural language.
- __Write Code__. Given prompting by the developer it can write code that you can one-click deploy into existing or new files.
- __Debug__. Copilot can analyze your code and propose solutions to fix bugs.

It can do most of what Labs can do with "brushes" as "topics", but whereas Labs changes the code in your file, the chat functionality just shows what it would change in the window. However, there is also an "inline mode" for GitHub Copilot Chat that allows you to make changes to your code inline which does not have this same limitation.

> You must sign-up for the preview [here](https://github.com/features/preview/copilot-x). Once you are accepted, you will be able to install the extension from the VS Code Marketplace. You may require the Insiders build of VS Code.

Author: removed.

> ## ChatGPT / Bing Chat

Reviewer: I would consider focusing this page on copilots and create another content for ChatGPT/Bing Chat.

Reviewer 2: I agree on focusing the page on copilots... 
Regarding placing ChatGPT/Bing Chat elsewhere, @shiranr do you feel that content fits in the EF playbook, or did you mean perhaps elsewhere as in not in the playbook?

Reviewer: This specific content is not relevant to the EF playbook but if people are using ChatGPT / Bing Chat for programming and can give a solid to do/ not to do about it, it MIGHT be good content for the playbook

> Prior to GitHub Copilot X chat availability, we used ChatGPT and Bing Chat to provide similar functionality. We found the generation of code using ChatGPT to be most valuable for these scenarios:

Reviewer 2: Same comment as before about wording, change the we... to some scenarios are... or something to that effect

> Prior to GitHub Copilot X chat availability, we used ChatGPT and Bing Chat to provide similar functionality. We found the generation of code using ChatGPT to be most valuable for these scenarios:

Author: For one, not everyone in our crew has been granted access to Copilot X yet. Secondly, Copilot will only answer "questions about programming", sometimes legitimate questions seem to be caught by that filter and we still use ChatGPT or Bing Chat for that.

I will add this context.

> Prior to GitHub Copilot X chat availability, we used ChatGPT and Bing Chat to provide similar functionality. We found the generation of code using ChatGPT to be most valuable for these scenarios:

Reviewer 2: As long as we keep it consistent with the topic so that it makes sense in the context of dev experience that is cool. We don't want it to scope creep and turn into a page about all things AI at Microsoft.

- __Write Code__. When using a library or framework that you are not familiar with, ChatGPT can help you get started by generating code.
- __Build SQL Indexes__. Given a query, ChatGPT can generate a SQL index that will improve the performance of the query.
- __Write Regular Expressions__. These are notoriously difficult to write, but ChatGPT can generate them for you if you give some sample input and describe what you want to extract.
- __Improve and Validate__. If you are unsure of the implications of writing code a particular way, you can ask ChatGPT questions about it. For instance, you might ask if there is a way to write the code that is more performant or uses less memory. Once it gives you an opinion, you can ask it to provide documentation validating that assertion.
- __Change Perspective__. ChatGPT can impersonate a persona or even a system and answer questions from that perspective. For example, you can ask it to explain what a particular piece of code does from the perspective of a user. You might have ChatGPT imagine it is a database adminstrator and ask it to explain how to improve a particular query.

> When using Bing Chat, we found much better results using Creative Mode.

Reviewer 2: ```suggestion
When using Bing Chat, experiment with modes, sometimes changing to creative mode can give the results you need.
```

Reviewer 2: We generally try to avoid blog/story telling wording like "our dev crew bla bla..." to keep a consistent tone of voice, also note that the playbook, while written by dev crews is meant both for our crews, customers and partners, so we should keep our internal structures out of the playbook and use more imperative language.

## Prompt Engineering

> Prompt Engineering is a cutting edge subject area focused on prompting strategies designed to derive the most value from an AI response.
> 
> Most LLM services such as GPT-4 or Bing Chat have a limited number of prompts a user can send to it in a given amount of time. Even outside of that scenario, we want to get the most useful information as quickly as possible.
> 
> This is one open source example of documentation around prompt engineering, and should be seen as a jumping off point for further investigation. [Link](https://github.com/brexhq/prompt-engineering)

Reviewer 2: ```suggestion
To get the best results with the copilot tools, a good prompt is crucial.

This repo can help serve as a starting point for learning and experimenting with prompt engineering. [Link](https://github.com/brexhq/prompt-engineering)
```

On first read, it was hard to follow how this related to the rest of the text or to dev experience. 

Suggestion: add context - see example above, and perhaps remove some of of the details about prompt engineering, that while true, are not necessary for the given content. We want to keep the playbook succinct and to the point, so that it is easy to find the content you need.

Author: I made it less verbose.

## Considerations

> It is important when using AI tools to understand how the data (including private or commercial code) might be used by the system. GitHub Copilot provides guarantees on how data will be used that should align with customer expectations, however, we should always seek a customer's consent. More information on the privacy can be found [here](https://resources.github.com/copilot-for-business/).

Reviewer 2: Again here, link to the page, and let them do the guarantees. If we make statements around guarantees on this page, we bind ourselves to update this if it ever changes. Else we will have two conflicting Microsoft sources on the topic.

Author: This doesn't say what those guarantees are.

One problem we had with ChatGPT is we couldn't provide it any prompts that included sensitive code. However, Copilot is designed to handle that scenario. Can you provide another way to say that?

> ChatGPT allows data entered into the prompts to be used for training, so private or commercial code should not be copied into that system.

Reviewer 2: Same as above...
